File: web/src/lib/avatar.ts
/**
 * Driving the dynamic-avatar-drawer from the game's traits.
 *
 * The library (vendored unmodified under public/vendor, LGPL v3) draws a stack
 * of canvases into a DOM element. This module owns two things and nothing
 * else: loading it, and the mapping from our traits to its dimensions. Its
 * dimensions are in its own units and ranges; ours are in centimetres and cup
 * sizes. Everything that translates between the two lives here, so neither
 * model has to know about the other.
 */

import { BUST, PHALLUS, type Scale } from "./traits/scale";
import type { Character } from "./traits/character";
import { NUDE } from "./traits/perception";

/** The library's own canvas size. The view scales this to fit its box. */
export const NATIVE_SIZE = [700, 1200] as const;

/**
 * `fem` is the library's core stat: "how overall feminine their appearance
 * is; influences a lot of dimensions". Range 0-11, average 5.
 */
const FEM_LOW = 0;
const FEM_HIGH = 11;

const SCRIPT_URL = "vendor/dynamic-avatar-drawer/da.js";

export interface AvatarPayload {
  readonly name: string;
  readonly fem: number;
  readonly basedim: Record<string, number>;
}

// -- the mapping ----------------------------------------------------------

/** Put a trait on one of the library's ranges, proportionally. */
function scaled(value: number, scale: Scale, low: number, high: number): number {
  const span = scale.maximum - scale.minimum;
  if (span <= 0) return low;
  const fraction = (value - scale.minimum) / span;
  return low + Math.max(0, Math.min(1, fraction)) * (high - low);
}

/**
 * The library's `fem` stat, from how the character reads.
 *
 * A happy coincidence: their core stat and our perception model are the same
 * idea. Ours runs -1 to +1 and already accounts for build, so it maps straight
 * onto their 0-11 without inventing anything. Read undressed, because this is
 * the body the drawing shows.
 */
export function femininity(character: Character): number {
  const score = character.presentation(NUDE).score; // -1 .. +1
  return FEM_LOW + ((score + 1) / 2) * (FEM_HIGH - FEM_LOW);
}

/** RGB 0-255 to hue (0-360), saturation and lightness (0-100). */
export function rgbToHsl(
  rgb: readonly [number, number, number],
): [number, number, number] {
  const [r, g, b] = [rgb[0] / 255, rgb[1] / 255, rgb[2] / 255];
  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  const lightness = (max + min) / 2;
  const delta = max - min;
  if (delta === 0) return [0, 0, lightness * 100];

  const saturation = delta / (1 - Math.abs(2 * lightness - 1));
  let hue: number;
  if (max === r) hue = ((g - b) / delta) % 6;
  else if (max === g) hue = (b - r) / delta + 2;
  else hue = (r - g) / delta + 4;
  hue *= 60;
  if (hue < 0) hue += 360;
  return [hue, saturation * 100, lightness * 100];
}

/**
 * Every library dimension we currently have an opinion about.
 *
 * Deliberately partial. The library has thirty-odd dimensions and we have six
 * traits; anything not named here keeps the library's own default, which is a
 * sensible average rather than a zero.
 */
export function dimensions(character: Character): Record<string, number> {
  const traits = character.traits;
  const [hue, saturation, lightness] = rgbToHsl(traits.get("hair_colour").rgb);
  return {
    // Their height is in centimetres too, so this one is a straight copy.
    height: traits.maybe("height", 170),
    // Cup index onto their 0-100ish scale.
    breastSize: scaled(traits.maybe("bust", 0), BUST, 0, 100),
    penisSize: scaled(traits.maybe("phallus", 0), PHALLUS, 0, 100),
    hairHue: hue,
    hairSaturation: saturation,
    hairLightness: lightness,
  };
}

/** What the library needs to build a player. */
export function payload(character: Character): AvatarPayload {
  return {
    name: character.name,
    fem: femininity(character),
    basedim: dimensions(character),
  };
}

/**
 * What the drawing depends on, as a comparable string.
 *
 * Compared rather than subscribed to: a trait changed by a path that forgot to
 * notify cannot leave a stale drawing.
 */
export function signature(character: Character): string {
  const data = payload(character);
  const dims = Object.entries(data.basedim)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([key, value]) => `${key}=${value.toFixed(3)}`);
  return [data.name, data.fem.toFixed(3), ...dims].join("|");
}

// -- loading the library --------------------------------------------------

interface DaLibrary {
  load(): Promise<unknown>;
  getCanvasGroup(id: string, size: { width: number; height: number }): HTMLElement;
  draw(group: HTMLElement, player: unknown, view: unknown): Promise<unknown>;
  Player: new (options: unknown) => unknown;
}

declare global {
  // eslint-disable-next-line no-var
  var da: DaLibrary | undefined;
}

let loading: Promise<DaLibrary> | null = null;

/**
 * Load and initialise the library, once per page.
 *
 * Injected rather than imported: it is a plain browser script that assigns a
 * global, and keeping it out of the bundle is also what keeps it replaceable,
 * which the LGPL requires of us.
 */
export function loadAvatarLibrary(): Promise<DaLibrary> {
  if (loading) return loading;
  loading = new Promise<DaLibrary>((resolve, reject) => {
    if (typeof document === "undefined") {
      reject(new Error("no document"));
      return;
    }
    if (globalThis.da) {
      resolve(globalThis.da);
      return;
    }
    const script = document.createElement("script");
    script.src = new URL(SCRIPT_URL, document.baseURI).toString();
    script.onload = () => {
      const library = globalThis.da;
      if (!library) {
        reject(new Error("da.js loaded but defined nothing"));
        return;
      }
      library.load().then(() => resolve(library), reject);
    };
    script.onerror = () => reject(new Error("da.js failed to load"));
    document.head.appendChild(script);
  });
  return loading;
}

/** The view options the game draws with; it prints its own name and stats. */
export const VIEW = {
  transparentBackground: true,
  printAdditionalInfo: false,
  printHeight: false,
  printVitals: false,
  renderShoeSideView: false,
  offsetX: 0,
  offsetY: 0,
} as const;

/**
 * One mounted avatar: a canvas group inside `host`, redrawn on demand.
 *
 * Draws are coalesced. A slider being dragged asks far faster than the library
 * can draw, and queueing every request would run the whole backlog after the
 * player stopped moving.
 */
export class AvatarView {
  private group: HTMLElement | null = null;
  private library: DaLibrary | null = null;
  private drawing = false;
  private again = false;
  private pending: AvatarPayload | null = null;
  private lastSignature: string | null = null;
  /** One line for what the library is doing, or why it is not drawing. */
  status = "loading";
  onStatus: ((status: string) => void) | null = null;

  constructor(private readonly host: HTMLElement) {}

  async mount(): Promise<void> {
    try {
      this.library = await loadAvatarLibrary();
    } catch (error) {
      this.setStatus(`unavailable: ${(error as Error).message}`);
      return;
    }
    // getCanvasGroup looks its holder up by id and reads its style without
    // checking -- its docstring claims it creates one, but it only creates
    // the canvases inside. Passing an id that does not exist throws on null.
    const holder = document.createElement("div");
    holder.id = `lechery-avatar-${Math.random().toString(36).slice(2)}`;
    this.host.appendChild(holder);
    this.group = this.library.getCanvasGroup(holder.id, {
      width: NATIVE_SIZE[0],
      height: NATIVE_SIZE[1],
    });
    this.setStatus("ready");
    if (this.pending) {
      const pending = this.pending;
      this.pending = null;
      this.draw(pending);
    }
  }

  /** Redraw for `character` if anything the drawing depends on moved. */
  update(character: Character, force = false): boolean {
    const current = signature(character);
    if (!force && current === this.lastSignature) return false;
    this.lastSignature = current;
    this.draw(payload(character));
    return true;
  }

  private draw(data: AvatarPayload): void {
    if (!this.library || !this.group) {
      this.pending = data;
      return;
    }
    if (this.drawing) {
      this.pending = data;
      this.again = true;
      return;
    }
    this.drawing = true;
    const player = new this.library.Player(data);
    this.library
      .draw(this.group, player, VIEW)
      .then(() => {
        this.drawing = false;
        if (this.again) {
          this.again = false;
          const next = this.pending;
          this.pending = null;
          if (next) this.draw(next);
        }
      })
      .catch((error: Error) => {
        this.drawing = false;
        this.setStatus(`draw failed: ${error.message}`);
      });
  }

  private setStatus(status: string): void {
    this.status = status;
    this.onStatus?.(status);
  }

  destroy(): void {
    this.group?.remove();
    this.group = null;
  }
}

File: lechery/ui/avatar.py
"""Driving the dynamic-avatar-drawer from the game's traits.

The library is a browser thing: it draws a stack of canvases into a DOM
element. So the drawn avatar exists only in the web build, and the
placeholder figure stays for the desktop. That is a real cost and worth
naming -- appearance is now something the desktop build cannot show.

This module owns the mapping from our traits to the library's dimensions,
which is the only interesting part. Its dimensions are in its own units and
its own ranges; ours are in centimetres and cup sizes. Everything that
translates between the two lives here so that neither model has to know
about the other.
"""

from __future__ import annotations

import colorsys
from typing import Optional

from ..traits import Character
from ..traits.scale import BUST, HEIGHT, PHALLUS
from . import dom

#: The library's own canvas size. Ours is scaled to fit whatever box the
#: paperdoll panel gives it.
NATIVE_SIZE = (700, 1200)

#: `fem` is the library's core stat: "how overall feminine their appearance
#: is; influences a lot of dimensions". Range 0-11, average 5.
FEM_LOW, FEM_HIGH = 0.0, 11.0


def available() -> bool:
    """Whether the drawn avatar can be used at all."""
    return dom.in_browser()


def ready() -> bool:
    """Whether the library has finished loading and drawn something."""
    return dom.evaluate(
        "window.LecheryAvatar ? String(window.LecheryAvatar.ready()) : 'false'"
    ) == "true"


def status() -> str:
    """One word for what the library is doing, or why it is not.

    There is no console on a phone, so a failure in the page is otherwise
    completely silent -- the panel draws this instead of an empty box.
    """
    reported = dom.evaluate(
        # Reports what is actually in the page, so a missing script and a
        # library that failed to load are not the same word.
        "window.LecheryAvatar ? window.LecheryAvatar.status()"
        " : ('no bridge (da=' + (typeof da) + ')')"
    )
    if reported is None:
        return "no browser"
    return reported


# -- the mapping ----------------------------------------------------------


def _scaled(value: float, scale, low: float, high: float) -> float:
    """Put a trait on one of the library's ranges, proportionally."""
    span = scale.maximum - scale.minimum
    if span <= 0:
        return low
    fraction = (float(value) - scale.minimum) / span
    return low + max(0.0, min(1.0, fraction)) * (high - low)


def femininity(character: Character) -> float:
    """The library's `fem` stat, from how the character reads.

    A happy coincidence: their core stat and our perception model are the
    same idea. Ours runs -1 to +1 and already accounts for build, so it
    maps straight onto their 0-11 without inventing anything.

    Read undressed, because this is the body the drawing shows.
    """
    from ..traits.perception import NUDE

    score = character.presentation(NUDE).score  # -1 .. +1
    return FEM_LOW + (score + 1.0) / 2.0 * (FEM_HIGH - FEM_LOW)


def hair_hsl(character: Character) -> tuple[float, float, float]:
    """Hair colour as the library wants it: hue, saturation, lightness.

    Our palette carries RGB, since that is what the placeholder figure
    draws with. Converting here keeps one source of truth for what
    "copper" means rather than a second colour table in another format.
    """
    colour = character.traits.get("hair_colour")
    rgb = getattr(colour, "rgb", (90, 70, 60))
    hue, lightness, saturation = colorsys.rgb_to_hls(*[c / 255 for c in rgb])
    return (hue * 360.0, saturation * 100.0, lightness * 100.0)


def dimensions(character: Character) -> dict[str, float]:
    """Every library dimension we currently have an opinion about.

    Deliberately partial. The library has thirty-odd dimensions and we have
    six traits; anything not named here keeps the library's own default,
    which is a sensible average rather than a zero.
    """
    traits = character.traits
    hue, saturation, lightness = hair_hsl(character)

    return {
        # Their height is in centimetres too, so this one is a straight copy.
        "height": float(traits.get("height", 170)),
        # Cup index onto their 0-100ish scale.
        "breastSize": _scaled(traits.get("bust", 0), BUST, 0.0, 100.0),
        "penisSize": _scaled(traits.get("phallus", 0), PHALLUS, 0.0, 100.0),
        "hairHue": hue,
        "hairSaturation": saturation,
        "hairLightness": lightness,
    }


def payload(character: Character) -> dict:
    """What the shim needs to build a player."""
    return {
        "name": character.name,
        "fem": femininity(character),
        "basedim": dimensions(character),
    }


def signature(character: Character) -> tuple:
    """What the drawing depends on, for deciding when to redraw.

    Compared rather than subscribed to, for the same reason the placeholder
    figure does it: a trait changed by a path that forgot to notify cannot
    leave a stale drawing.
    """
    data = payload(character)
    return (data["name"], round(data["fem"], 3)) + tuple(
        round(value, 3) for _, value in sorted(data["basedim"].items())
    )


# -- driving the page -----------------------------------------------------


class Avatar:
    """The drawn figure, positioned over a rect the panel chooses.

    It lives in the page, not on the game's canvas, so it does not
    disappear just because the thing that placed it stopped drawing. The
    frame protocol handles that: whoever wants it visible must say so every
    frame, and `frame_done` takes it away when nobody did.
    """

    def __init__(self) -> None:
        self._signature: Optional[tuple] = None
        self._rect = None
        self.visible = False
        self._placed_this_frame = False

    def update(self, character: Character) -> bool:
        """Redraw if the character changed. Returns whether it did."""
        current = signature(character)
        if current == self._signature:
            return False
        self._signature = current
        dom.call_json("LecheryAvatar.update", payload(character))
        return True

    def place(self, rect) -> None:
        """Show the avatar over `rect`, given in device pixels."""
        self._rect = rect
        self._placed_this_frame = True
        if rect == self._rect and self.visible:
            return
        left, top, width, height = dom.css_geometry(rect)
        dom.call("LecheryAvatar.place", left, top, width, height)
        self.visible = True

    def hide(self) -> None:
        if not self.visible:
            return
        dom.call("LecheryAvatar.hide")
        self.visible = False

    def frame_done(self) -> None:
        """End of frame: hide unless something asked for it.

        Without this the figure stays over whatever replaced the screen
        that was showing it -- the menu, most obviously, which does not
        draw the paperdoll panel at all.
        """
        if not self._placed_this_frame:
            self.hide()
        self._placed_this_frame = False


#: One avatar, because there is one element in the page.
_SHARED = Avatar()


def shared() -> Avatar:
    return _SHARED


def frame_done() -> None:
    _SHARED.frame_done()
